### ISSUE TITLE
Accessibility work in OpenRefine's UI

### ISSUE THREAD

**Issue by wetneb** [wetneb/hello-elebitzero#2]
Hi @elebitzero!

OpenRefine currently has [a grant to improve diversity and inclusion in the project](https://chanzuckerberg.com/eoss/proposals/openrefine-for-everyone/), and we want to use part of that to work on the accessibility of the UI.

You have recently been active in that area, for instance adding missing `<label>` and `aria` in the UI, improving the contrast of the histogram of the numerical facets, and so on. Therefore I would like your opinion on how we could support that effort.

One thing we are thinking about is to commission an accessibility review of the UI. There are plenty of agencies which offer those services, producing a report explaining in which ways a website or application differs from the [WCAG](https://www.w3.org/WAI/standards-guidelines/wcag/) for instance.
We could then break down such a report into actionable GitHub issues, and potentially fund work to tackle some of them.

Alternatively, I can imagine you have already identified some accessibility problems, and perhaps commissioning such a review can feel wasteful if we expect to get told things we were already aware of before. So I think it should also be possible to fund accessibility fixes straight away, without having to first go through an external review (which could still be done later on, when we think we have covered all issues we are aware of).

Let me know what you think about this! Also, if you think it is worth getting an external review, and if you have agencies to recommend for that, I would be interested.

Cheers!

**Comment by elebitzero** [comment]
There is a lot of accessibility work that could be done. Most of these are just following the latest best practices. There are also tools which can point out problems.

Regarding text contrast, I also corrected the text background to increase text contrast on the preferences page.  There are a few places that need work, but the text contrast is good in most places.

I don't use keyboard navigation much, but noticed there are lots of problems.  On the index page, tabbing won't tab between the tabs 'Create Project', 'Import Project', etc. On the project page, tabbing won't select the column dropdown menus. If you have a modal open, once tabbing cycles through the dialog's elements, it starts cycling through the main console.

I have added missing labels and aria-labels. There are still some missing, especially those which require dynamic labels, like the 'Field XXX Allow null' for checkboxes on Export SQL dialog.

Also links and buttons which use images/SVG instead of text need aria-labels e.g. for X icon, flag, star links. Some of the images don't have alt text.

OpenRefine's HTML is using HTML tables for the layout even when the content is not a table, which is causing problems with screen readers. You can open a dialog with two input controls and the screen reader will say 'Table with two rows. Row 1 Col 1....' which is a major issue. I have updated the layout on a few dialogs as I worked on them, and a fair amount have this problem.  Everything needs to be changed to semantic-based layout.

**Comment by wetneb** [comment]
Ok, thanks a lot for this very insightful overview!

I was not aware that tables were an accessibility concern, I thought they were just frowned upon for other reasons. I have to admit I still have the tendency to use them from time to time, so this is a useful heads up!

Although you are already aware of many issues already, it does feel like an external review could be helpful, for the sake of it being systematic and independent, no? Perhaps there is some way to prioritize those issues, too. At least I would find it useful to raise awareness of the problems in OpenRefine's committees.

The broader question I have on my mind is, to what extent should we work on fixing those accessibility issues with the current web framework? I assume that if we were to migrate to some pre-existing UI component library (and followed its intended usage) then we would already be solving a lot of such issues in one go. But surely not all. And I am not saying I have plans to initiate such a migration soon either! And I guess that is the sort of question you can ask for just about any aspect of OpenRefine (is it worth renovating this part when the rest is so crumbling?), but apparently we are still here, so perhaps it's not worth getting too hung up in those concerns.
